lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -569,6 +576,99 @@ newTransactionLayer networkId = TransactionLayer
 _decodeSealedTx :: SealedTx -> (Tx, TokenMap, TokenMap, [Certificate])
 _decodeSealedTx (cardanoTx -> InAnyCardanoEra _era tx) = fromCardanoTx tx
 
+_evaluateTransactionBalance
+    :: SealedTx
+    -> Cardano.ProtocolParameters
+    -> UTxO
+    -> [(TxIn, TxOut, Maybe (Hash "Datum"))]
+    -> Maybe Cardano.Value
+_evaluateTransactionBalance tx pp utxo extraUTxO = do
+    shelleyTx <- inAnyShelleyBasedEra $ cardanoTx tx
+    pure $ withShelleyBasedBody shelleyTx $ \era bod ->
+        let
+            utxo' = Map.fromList
+                . map (bimap toCardanoTxIn (toCardanoTxOut era))
+                . Map.toList
+                $ unUTxO utxo
+
+            extraUTxO' = Map.fromList
+                . map (\(i, o, mDatumHash) ->
+                    (toCardanoTxIn i
+                    , setDatumHash era mDatumHash (toCardanoTxOut era o))
+                    )
+                $ extraUTxO
+
+        in
+            lovelaceFromCardanoTxOutValue
+                $ Cardano.evaluateTransactionBalance
+                    pp
+                    mempty
+                    (Cardano.UTxO $ utxo' <> extraUTxO')
```

Comment:
Perhaps we could add a comment to say why the order of parameters around `<>` is important?

I'm guessing that values in `extraUTxO` are expected to override those in `utxo'`?

Reply:
2453aa1fc0e79a0978deb501138ff8b9937243fc